File: digitalmedia/Providers/YouTubeProvider.php
<?php namespace DigitalMedia\Providers;

use DigitalMedia\Interfaces\DigitalMediaProvider;
use DigitalMedia\Models\YouTubeModel;
use DigitalMedia\Utility\GoogleClient;
use GuzzleHttp\Client;
use Carbon\Carbon;

class YouTubeProvider implements DigitalMediaProvider
{
  const PUBLISHER_KEY = "youtube";

  protected $client = null;
  private $channel_id;

  function __construct($channel_id, $api_key)
  {
    $this->client = new GoogleClient($api_key);
    $this->channel_id = $channel_id;
  }

  public function getPublishedContent($count)
  {
    $service = new \Google_Service_YouTube($this->client->gclient);

    $search_array = array(
      'channelId' => $this->channel_id,
      'type' => 'video'
    );

    if(!is_null($count))
    {
      $search_array['maxResults'] = $count;
    }

    $response = $service->search->listSearch('id,snippet', $search_array);

    $video_list = array();
    $youtube_video_format = 'https://www.youtube.com/watch?v=%s';

    foreach($response['items'] as $item)
    {
      $model = new YouTubeModel(self::PUBLISHER_KEY);
      $model->id = $item['id']['videoId'];
      $model->date_created = new Carbon($item['snippet']['publishedAt']);
      $model->title = $item['snippet']['title'];
      $model->description = $item['snippet']['description'];
      $model->url = sprintf($youtube_video_format, $model->id);
      $model->thumbnail = $item['snippet']['thumbnails']['high']['url'];

      array_push($video_list, $model);
    }

    return $video_list;
  }

  public function getPublisherKey(){
    return self::PUBLISHER_KEY;
  }

  public function getPublicationType()
  {
    return YouTubeModel::TYPE_NAME;
  }
}

File: lib/media/parsers/youtube_provider.rb
require 'Nokogiri'
require 'google/apis/youtube_v3'
require_relative '../../media/entities/youtube'

module Nads
  module Media
    module Providers
      class YoutubeProvider
        def initialize(developer_key, channel_id)
          @key = developer_key
          @channel = channel_id
          @service_name = 'youtube'
          @api_version = 'v3'
        end

        # @params[Integer] count
        # @returns[Array]
        def get_latest(count=10)
          youtube = Google::Apis::YoutubeV3

          client = youtube::YouTubeService.new

          client.key = @key

          resp = client.list_searches('id,snippet', channel_id: @channel, type: 'video', max_results: count, order: :date)

          entities = []

          resp.items.each do |item|
            entity = Nads::Media::Entities::YouTube.new

            entity.id = item.id.video_id
            entity.date_created = item.snippet.published_at
            entity.title = item.snippet.title
            entity.description = item.snippet.description
            entity.url = "https://www.youtube.com/watch?v=#{entity.id}"
            entity.thumbnail = item.snippet.thumbnails.high.url

            entities.push(entity)
          end

          resp
        end
      end
    end
  end
end